### PR TITLE
Add ability to build a subset of things

### PR DESCRIPTION
**Pull request by mook-as**
This allows you to build only the packages required for a named selection of roles, or their images, instead of doing everything at once.

This should make it easier to CI things.

I'm introducing sub-vipers to work around issues where it gets confused if multiple subcommands use the same flag name.

### CODE REVIEW

File: app/fissile_test.go [hunk 1]
```diff
@@ -177,3 +177,60 @@ func TestDevDiffConfigurations(t *testing.T) {
 	v, ok = hashDiffs.ChangedValues["cf.bogus.key"]
 	assert.False(ok)
 }
+
+func TestFisileSelectRolesToBuild(t *testing.T) {
```
Comment (andreas-kupries) [inline app/fissile_test.go]: Fisile => Fissile (s vs ss)

File: cmd/build-packages.go [hunk 2]
```diff
@@ -25,6 +28,8 @@ compiled once.
 `,
 	RunE: func(cmd *cobra.Command, args []string) error {
 
+		flagBuildPackagesRoles := viper.GetString("roles")
```
Comment (mook-as) [inline cmd/build-packages.go]: I know this works (at least, I've tested it), but I have no idea _how_.  This is talking to the global viper, not the custom one on line 55...

File: app/fissile.go [hunk 3]
```diff
@@ -444,8 +449,35 @@ func (f *Fissile) GeneratePackagesRoleImage(repository string, roleManifest *mod
...
+
+	if len(roleNames) == 0 {
+		// No role names specified, assume all roles
+		return roles, nil
+	}
+
+	var missingRoles []string
+	roleMap := make(map[string]*model.Role, len(roles))
+	for _, role := range roles {
+		roleMap[strings.ToLower(role.Name)] = role
```
Comment (viovanov) [inline app/fissile.go]: iirc our object model already has a map for role name to role

Comment (viovanov) [inline app/fissile.go]: yeah - https://github.com/hpcloud/fissile/blob/master/model/roles.go#L42

File: app/fissile.go [hunk 4]
```diff
@@ -444,8 +449,35 @@ func (f *Fissile) GeneratePackagesRoleImage(repository string, roleManifest *mod
 	return nil
 }
 
+func (f *Fissile) selectRolesToBuild(roles model.Roles, roleNames []string) (model.Roles, error) {
```
Comment (viovanov) [inline app/fissile.go]: This function can be part of `RoleManifest`

File: builder/packages_image.go [hunk 5]
```diff
@@ -136,17 +136,17 @@ func (p *PackagesImageBuilder) determinePackagesLayerBaseImage(packages model.Pa
 }
 
 // NewDockerPopulator returns a function which can populate a tar stream with the docker context to build the packages layer image with
-func (p *PackagesImageBuilder) NewDockerPopulator(roleManifest *model.RoleManifest, forceBuildAll bool) func(*tar.Writer) error {
+func (p *PackagesImageBuilder) NewDockerPopulator(roles model.Roles, forceBuildAll bool) func(*tar.Writer) error {
```
Comment (viovanov) [inline builder/packages_image.go]: hmm - so the built subset would have its own packages layer?
I don't think that's what we want

File: cmd/root.go [hunk 6]
```diff
@@ -170,20 +170,25 @@ func init() {
 
 // initConfig reads in config file and ENV variables if set.
 func initConfig() {
+	initViper(viper.GetViper())
+}
+func initViper(v *viper.Viper) {
 	if cfgFile != "" { // enable ability to specify config file via flag
-		viper.SetConfigFile(cfgFile)
```
Comment (viovanov) [inline cmd/root.go]: What's the reason for these changes?

Comment (mook-as) [inline cmd/root.go]: So that I can have multiple commands that have a `--role` argument. Otherwise it gets confused, and always uses empty for one of them (the first one bound, I think; binding overrides any existing arguments with the same name).

File: builder/packages_image.go (hunk 5 above)
Comment (mook-as) [inline builder/packages_image.go]: It needs to, otherwise we'd try to build all the packages even for one role (e.g. if we wanted gorouter, we don't need Java for UAA).

That _should_ be okay, since (for development) it should still build them based on existing package layers if they already contain the correct packages (via docker labels).

Comment (mook-as) [inline builder/packages_image.go]: Wait what we don't actually have that code. Let me put that in...

File: cmd/root.go (hunk 6 above)
Comment (mook-as) [inline cmd/root.go]: Ah, we _do_ have that code, near https://github.com/mook-as/fissile/blob/ebd1ad311ea40b474688182d5ce9babd88d163e0/builder/packages_image.go#L165
